### PR TITLE
Pcawg2-dkfz

### PR DESCRIPTION
**Pull request by lindaxiang**
1. Refactor the Dockerfile and build an image which can get the dkfz workflow runnable.
2. We test and compare the results using real pcawg-1 data and verify the results are good
3. Make modifications to take care of ticket: https://github.com/ICGC-TCGA-PanCancer/dkfz_dockered_workflows/issues/22

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -3,134 +3,153 @@ MAINTAINER Michael Heinold @ DKFZ
...
-
-RUN pip install --index-url=https://pypi.python.org/simple/ pip==10.0.1
-RUN pip install scipy==0.12.0
 
+RUN echo `head -n 1 /etc/hosts | cut -f1` master > /tmp/hostsTmp \
+    && tail -n +2 /etc/hosts >> /tmp/hostsTmp \
+    && cp /tmp/hostsTmp /etc/hosts \
+    && echo master > /etc/hostname \
+	&& hostName=`hostname` \
+    && export HOST=$hostName \
```
Comment (junjun-zhang) [inline Dockerfile]: lines 74-79 repeat lines 55-60, it's totally unnecessary.

File: Dockerfile [hunk 2]
```diff
@@ -3,134 +3,153 @@ MAINTAINER Michael Heinold @ DKFZ
...
+
+COPY scripts/sgeInit.sh /roddy/sgeInit.sh
+
+RUN echo `head -n 1 /etc/hosts | cut -f 1` master > /tmp/hostsTmp \
+	&& tail -n +2 /etc/hosts >> /tmp/hostsTmp \
+	&& cp /tmp/hostsTmp /etc/hosts \
+	&& echo master > /etc/hostname \
+	&& cat /etc/hosts \
+	&& hostName=`hostname` \
+    && export HOST=$hostName \
```
Comment (junjun-zhang) [inline Dockerfile]: lines 90-96 repeat 55-60, it's totally unnecessary

File: Dockerfile [hunk 3]
```diff
@@ -3,134 +3,153 @@ MAINTAINER Michael Heinold @ DKFZ
...
-    apt-get install -y samtools python-apt python-yaml python-jinja2 git wget sudo;\
-    git clone http://github.com/ansible/ansible.git /ansible
+
 # get a specific version of ansible , add sudo to seqware, create a working directory
-RUN git checkout v1.6.10 ;
+RUN git clone http://github.com/ansible/ansible.git /ansible \
+    && git checkout v1.6.10
 ENV PATH /ansible/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
 ENV ANSIBLE_LIBRARY /ansible/library
 ENV PYTHONPATH /ansible/lib:$PYTHON_PATH
```
Comment (junjun-zhang) [inline Dockerfile]: is this supposed to be: `ENV PYTHONPATH /ansible/lib:$PYTHONPATH`?

**Review by junjun-zhang — COMMENTED** [review]
It looks good, left minor comments.

File: Dockerfile (hunk 1 above)
Comment (dankwart-de) [reply Dockerfile]: Hi all, 
as I worked on the initial container, I'll just give a comment, why we have these repeated hostname and gridengine steps. When we developed the container, we worked with the cache feature of Docker. When one uses this, at least in my experience, the hostname changes constantly and for SGE to work properly, the hostname MUST be set upon every RUN directive. Unfortunately I am not aware, if the current Docker files or build commands do not have this problem.

**Comment by dankwart-de** [comment]
I cannot accept this, as I cannot tell if it works, but nice work though! 👍

File: scripts/run_workflow.pl [hunk 4]
```diff
@@ -41,17 +41,17 @@
...
+run("ln -sf $bedpe /data/datastore/delly/delly.bedpe.txt");
 run("mkdir -p /mnt/datastore/workflow_data/");
 run("mkdir -p \$TMPDIR/reference");
 # make sure we have permissions on these volumes
 run("sudo chmod -R a+wrx /reference");
 run("cd \$TMPDIR/reference && tar zxf $reference");
-run("mkdir -p /mnt/datastore/ && ln -s \$TMPDIR/reference/bundledFiles /mnt/datastore/");
+run("mkdir -p /mnt/datastore/ && ln -sf \$TMPDIR/reference/bundledFiles /mnt/datastore/");
 run("sudo chmod -R a+wrx /mnt/datastore /data/datastore");
 run("mkdir -p /mnt/datastore/resultdata");
```
Comment (denis-yuen) [inline scripts/run_workflow.pl:57]: Warning: this isn't directly in scope, but line 81 is affected by 
https://github.com/common-workflow-language/common-workflow-language/issues/674
`/var/spool/cwl/` will need to be changed to `$(runtime.outdir)`  for compatibility with newer cwltool versions

I think there's also a reference in start.sh

**Review by denis-yuen — COMMENTED** [review]
A heads-up

**Comment by junjun-zhang** [comment]
@dankwart-de, our goal is to make it work and reproduce same results as in PCAWG production. The acceptance criteria is to get same result using real PCAWG aligned BAMs as input.

File: scripts/run_workflow.pl (hunk 4 above)
Comment (junjun-zhang) [reply scripts/run_workflow.pl:57]: Thanks for the heads-up, this indeed happens to other PCAWG workflows as well. As you mentioned it's out of scope for this fix and PR, we can address `/var/spool/cwl` related issues for all PCAWG workflows separately.

For now, we use an older version of `cwltool` to run the workflows. We will add to `doc` section the cwltool version used to test the workflow successfully.

**Review by junjun-zhang — APPROVED** [review]
Tested using real PCAWG BAMs, results match up with production run.

**Pull request merged**